src/components/chatbox/message.tsx

```diff
@@ -0,0 +1,317 @@
+import { useState, useEffect } from 'react'
+import classNames from 'classnames'
+import ReactMarkdown from 'react-markdown'
+import remarkGfm from 'remark-gfm'
+import rehypePrism from '@mapbox/rehype-prism'
+import { MdxPre } from 'components/dev-dot-content/mdx-components/mdx-code-blocks'
+
+// https://helios.hashicorp.design/icons/library
+
+import { IconCheckSquare24 } from '@hashicorp/flight-icons/svg-react/check-square-24'
+import { IconClipboard24 } from '@hashicorp/flight-icons/svg-react/clipboard-24'
+import { IconThumbsDown24 } from '@hashicorp/flight-icons/svg-react/thumbs-down-24'
+import { IconThumbsUp24 } from '@hashicorp/flight-icons/svg-react/thumbs-up-24'
+import { IconWand24 } from '@hashicorp/flight-icons/svg-react/wand-24'
+
+import Button from 'components/button'
+import Text from 'components/text'
+import FeedbackForm from 'components/feedback-form'
+import IconTile from 'components/icon-tile'
+import useAuthentication from 'hooks/use-authentication'
+
+import s from './message.module.css'
+
+const UserMessage = ({ text, image }: { text: string; image?: string }) => {
+	return (
+		<div className={classNames(s.message, s.messageUser)}>
+			<div className={classNames(s.avatar)}>
+				{image ? <img src={image} alt="user avatar" /> : null}
+			</div>
+			<Text /* Body/200/Medium */
+				size={200}
+				weight="medium"
+				className={classNames(s.content, s.userInput)}
+			>
+				{text}
+			</Text>
+			<div className={classNames(s.gutter)}></div>
+		</div>
+	)
+}
+
+const AssistantMessage = ({
+	markdown,
+	showActions,
+	conversationId,
+	messageId,
+}: {
+	markdown: string
+	showActions: boolean
+	conversationId: string
+	messageId: string
+}) => {
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+	// Determines green/red button
+	const [rating, setRating] = useState<1 | -1 | 0>(0)
+
+	const handleFeedback = async ({
+		rating,
+		text,
+	}: {
+		rating?: -1 | 1
+		text?: string
+	}) => {
+		const response = await fetch('/api/chat/feedback', {
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/json',
+				Authorization: `Bearer ${accessToken}`,
+			},
+			body: JSON.stringify({
+				messageId,
+				conversationId,
+				rating,
+				text,
+			}),
+		})
+		return response
+	}
+
+	const [isCopied, setIsCopied] = useState(false)
+	useEffect(() => {
+		let id: ReturnType<typeof setTimeout>
+		if (isCopied) {
+			id = setTimeout(() => {
+				setIsCopied(false)
+			}, 4000)
+		}
+		return () => {
+			clearTimeout(id)
+		}
+	}, [isCopied])
+
+	/** copy rich text; not raw markdown */
+	const handleCopy = async () => {
+		// CSS Module classname is a reasonable "hook" to grab the element
+		// rendered by ReactMarkdown.
+		const el = document.getElementsByClassName(s.markdown)?.[0]
+
+		if (el) {
+			// https://www.nikouusitalo.com/blog/why-isnt-clipboard-write-copying-my-richtext-html/
+			const clipboardItem = new ClipboardItem({
+				// @ts-expect-error - innerText is available
```

Comment:
⛏️ Nitpick: I think you can get rid of the `@ts-expect-error` here with a (somewhat wordy) type guard:

```suggestion
		if (el && 'innerText' in el && typeof el.innerText === 'string') {
			// https://www.nikouusitalo.com/blog/why-isnt-clipboard-write-copying-my-richtext-html/
			const clipboardItem = new ClipboardItem({
```

Reply:
That looks like it _should_ satisfy the compiler but it doesn't seem to work — going to skip this change

<img width="545" alt="CleanShot 2023-09-22 at 11 29 03@2x" src="https://github.com/hashicorp/dev-portal/assets/26389321/a66abbad-1c24-4f7a-abf9-673deedd6a49">